packages/muelu/src/Misc/MueLu_SegregatedAFactory_def.hpp

```diff
@@ -55,114 +55,273 @@
+    Teuchos::ArrayView<const Scalar> vals;
+    Ain->getLocalRowView(row, indices, vals);
 
-    // fetch map from level
-    RCP<Map> map = Teuchos::null;
-    if (currentLevel.GetLevelID() == 0) {
-      map = currentLevel.Get<RCP<Map>>(mapName, NoFactory::get());
-      GetOStream(Statistics0) << "User provided map \"" << mapName << "\": length dimension=" << map->getGlobalNumElements() << std::endl;
-    } else {
-      if (currentLevel.IsAvailable(mapName, mapFact_.get()) == false)
-        GetOStream(Runtime0) << "User provided map \"" << mapName << "\" not found in Level class on level " << currentLevel.GetLevelID() << "." << std::endl;
-      map = currentLevel.Get<RCP<Map>>(mapName, mapFact_.get());
-    }
+    // just copy all values in output
+    Teuchos::ArrayRCP<GlobalOrdinal> indout(indices.size(),Teuchos::ScalarTraits<GlobalOrdinal>::zero());
+    Teuchos::ArrayRCP<Scalar>        valout(indices.size(),Teuchos::ScalarTraits<Scalar>::zero());
+
+    size_t nNonzeros = 0;
+    for(size_t i=0; i<(size_t)indices.size(); i++) { // or can be parallelize this loop?
+      GlobalOrdinal gcid = Ain->getColMap()->getGlobalElement(indices[i]); // global column id
+      bool isInMap2 = dropMap1->isNodeGlobalElement(gcid);
 
-    // create new empty Operator
-    Teuchos::RCP<Matrix> Aout = MatrixFactory::Build(Ain->getRowMap(), Ain->getGlobalMaxNumRowEntries());
-
-    size_t numLocalRows = Ain->getLocalNumRows();
-    for(size_t row=0; row<numLocalRows; row++) { // how can i replace this by a parallel for?
-      GlobalOrdinal grid = Ain->getRowMap()->getGlobalElement(row); // global row id
-      bool isInMap = map->isNodeGlobalElement(grid);
-
-      // extract row information from input matrix
-      Teuchos::ArrayView<const LocalOrdinal> indices;
-      Teuchos::ArrayView<const Scalar> vals;
-      Ain->getLocalRowView(row, indices, vals);
-
-      // just copy all values in output
-      Teuchos::ArrayRCP<GlobalOrdinal> indout(indices.size(),Teuchos::ScalarTraits<GlobalOrdinal>::zero());
-      Teuchos::ArrayRCP<Scalar>        valout(indices.size(),Teuchos::ScalarTraits<Scalar>::zero());
-
-      size_t nNonzeros = 0;
-      for(size_t i=0; i<(size_t)indices.size(); i++) { // or can be parallelize this loop?
-        GlobalOrdinal gcid = Ain->getColMap()->getGlobalElement(indices[i]); // global column id
-        bool isInMap2 = map->isNodeGlobalElement(gcid);
-
-        if (isInMap == isInMap2) {
-          indout [nNonzeros] = gcid;
-          valout [nNonzeros] = vals[i];
-          nNonzeros++;
-        }
+      if (isInMap == isInMap2) {
+        indout [nNonzeros] = gcid;
+        valout [nNonzeros] = vals[i];
+        nNonzeros++;
       }
-      indout.resize(nNonzeros);
-      valout.resize(nNonzeros);
+    }
+    indout.resize(nNonzeros);
+    valout.resize(nNonzeros);
+
+    Aout->insertGlobalValues(Ain->getRowMap()->getGlobalElement(row), indout.view(0,indout.size()), valout.view(0,valout.size()));
```

Comment:
I don't think so. Since we provide GIDs in the dropMaps, we need to convert LIDs to GIDs for the check if entries need to be dropped. This is the information we work with and which we can then use to insert the values in the final matrix.